### PR TITLE
fix(npm): massage lockfile with rangeStrategy=update-lockfile

### LINKED ISSUE #14584: package-lock.json has incorrect modifications with rangeStrategy=update-lockfile
### How are you running Renovate?

WhiteSource Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

_No response_

### Please select which platform you are using if self-hosting.

_No response_

### If you're self-hosting Renovate, tell us what version of the platform you run.

_No response_

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the bug

We previously were using lockfile version 1 with npm 6 and I did not notice it there.

Since we migrated to lockfile version 2 with npm 8 I noticed that apparently Renovate uses e.g. `npm install --package-lock-only --no-audit --ignore-scripts postcss@8.4.8` to update postcss to 8.4.8. If there is a constraint like `^8.0.0` then npm updates also the `package.json` and relevant parts of it in `package-lock.json`.

Since I'm using `rangeStrategy=update-lockfile` only the lockfile is commited, which has still modifications that should not be there. When I do a local `npm install` they get reverted again, which is a bit annoying, e.g.
```diff
--- a/package-lock.json
+++ b/package-lock.json
@@ -23,7 +23,7 @@
         "css-minimizer-webpack-plugin": "^3.0.0",
         "mini-css-extract-plugin": "^2.0.0",
         "node-sass": "^7.0.0",
-        "postcss": "^8.4.8",
+        "postcss": "^8.0.0",
         "sass-loader": "^12.0.0",
         "terser-webpack-plugin": "^5.0.0",
         "vue-loader": "^15.5.1",
```

This is basically the follow-up of https://github.com/renovatebot/renovate/discussions/14562 where I initially was confused about that.

We're actually self-hosting Renovate and I noticed the problem with versions 31 and 32 on GitLab and Bitbucket. BUT I created a simple reproduction repo at https://github.com/herndlm/renovate-npm-lockfile-diff-repro that shows the problem.
While doing that I realized that this might be a `npm` limitation that is not easy to work around. Apparently it happens whenever the npm install command is called and I (a npm novice) couldn't figure out how to update the lockfile without updating package.json.
The following line should not have been changed https://github.com/herndlm/renovate-npm-lockfile-diff-repro/pull/3/files#diff-053150b640a7ce75eff69d1a22cae7f0f94ad64ce9a855db544dda0929316519R12

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Executing command(branch="renovate/postcss-8.x-lockfile")
{
  "command": "docker run --rm --name=renovate_node --label=renovate_child -v \"/mnt/renovate/gh/herndlm/renovate-npm-lockfile-diff-repro\":\"/mnt/renovate/gh/herndlm/renovate-npm-lockfile-diff-repro\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -e NPM_CONFIG_CACHE -e npm_config_store -w \"/mnt/renovate/gh/herndlm/renovate-npm-lockfile-diff-repro\" docker.io/renovate/node bash -l -c \"install-tool npm 8.5.3 && hash -d npm 2>/dev/null || true && npm install --package-lock-only --no-audit --ignore-scripts postcss@8.4.8\""
}
DEBUG: exec completed(branch="renovate/postcss-8.x-lockfile")
{
  "cmd": "docker run --rm --name=renovate_node --label=renovate_child -v \"/mnt/renovate/gh/herndlm/renovate-npm-lockfile-diff-repro\":\"/mnt/renovate/gh/herndlm/renovate-npm-lockfile-diff-repro\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -e NPM_CONFIG_CACHE -e npm_config_store -w \"/mnt/renovate/gh/herndlm/renovate-npm-lockfile-diff-repro\" docker.io/renovate/node bash -l -c \"install-tool npm 8.5.3 && hash -d npm 2>/dev/null || true && npm install --package-lock-only --no-audit --ignore-scripts postcss@8.4.8\"",
  "durationMs": 14150,
  "stdout": "Installing legacy tool npm v8.5.3\n\nadded 1 package, and audited 208 packages in 6s\n\n10 packages are looking for funding\n  run `npm fund` for details\n\nfound 0 vulnerabilities\n8.5.3\n\nup to date in 1s\n\n1 package is looking for funding\n  run `npm fund` for details\n",
  "stderr": "npm WARN using --force Recommended protections disabled.\nnpm notice \nnpm notice New minor version of npm available! 8.3.1 -> 8.5.3\nnpm notice Changelog: <https://github.com/npm/cli/releases/tag/v8.5.3>\nnpm notice Run `npm install -g npm@8.5.3` to update!\nnpm notice \nnpm WARN using --force Recommended protections disabled.\nchmod: cannot access '/home/ubuntu/.npm': No such file or directory\n"
}
DEBUG: package-lock.json needs updating(branch="renovate/postcss-8.x-lockfile")
DEBUG: Updated 1 lock files(branch="renovate/postcss-8.x-lockfile")
{
  "updatedArtifacts": [
    "package-lock.json"
  ]
}
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

### PR DESCRIPTION
**Pull request by herndlm**
## Changes

Fixes modifications in `package-json.lock` in lockfileVersion 2 when using `rangeStrategy=update-lockfile` that exist because npm install is called with a specific version.

The parsing logic is heavily based on https://github.com/renovatebot/renovate/blob/32.0.6/lib/modules/manager/npm/post-update/index.ts#L174

- is this the best place and way to do that?
- ~~if and how to test this with a unit test? guess I need a fixture? I can do that, just need a hint of how it should be done please :)~~
- ~~does it also work with lockfile version 1? I think it's not needed there, so maybe it should be limited to 2~~

## Context

Closes https://github.com/renovatebot/renovate/issues/14584

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/npm/post-update/npm.ts [hunk 1]
```diff
@@ -123,6 +124,26 @@ export async function generateLockFile(
...
+    if (lockUpdates.length) {
+      let detectedIndent: string;
+      let lockFileParsed: any;
+      try {
+        detectedIndent = detectIndent(lockFile).indent || '  ';
+        lockFileParsed = JSON.parse(lockFile);
+      } catch (err) {
+        logger.warn({ err }, 'Error parsing npm lock file');
+      }
+      if (lockFileParsed && lockFileParsed?.lockfileVersion === 2) {
```
Comment (rarkins) [inline lib/modules/manager/npm/post-update/npm.ts:139]: I've never looked closely but apparently there's a version 3 as well. Not sure if now is the right time to take it into account though

File: lib/modules/manager/npm/post-update/npm.ts [hunk 2]
```diff
@@ -123,6 +124,26 @@ export async function generateLockFile(
...
+        lockUpdates.forEach((lockUpdate) => {
+          lockFileParsed.packages[''][lockUpdate.depType][lockUpdate.depName] =
```
Comment (rarkins) [inline lib/modules/manager/npm/post-update/npm.ts:141]: We need to add a `if (lockFileParsed.packages?.['']?.[lockUpdate.depType]?.[lockUpdate.depName])` defensive check first

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
@herndlm do you have any public repo you can demo this against, when done?

File: lib/modules/manager/npm/post-update/npm.ts (hunk 1 above)
Comment (herndlm) [reply lib/modules/manager/npm/post-update/npm.ts:139]: very good point. I checked that, summary: it should be fine like that.

longer description:
According to https://docs.npmjs.com/cli/v7/configuring-npm/package-lock-json#lockfileversion, there should be a version 3 lock file at `node_modules/.package-lock.json`. I used my reproduction repo and compared the 2 lockfiles, the diff is
```diff
--- package-lock.json	2022-03-13 14:00:40.000000000 +0100
+++ node_modules/.package-lock.json	2022-03-13 14:00:40.000000000 +0100
@@ -4,14 +4,6 @@
   "lockfileVersion": 2,
   "requires": true,
   "packages": {
-    "": {
-      "name": "renovate-npm-lockfile-diff-repro",
-      "version": "1.0.0",
-      "license": "ISC",
-      "dependencies": {
-        "postcss": "^8.0.0"
-      }
-    },
     "node_modules/nanoid": {
       "version": "3.3.1",
       "resolved": "https://registry.npmjs.org/nanoid/-/nanoid-3.3.1.tgz",
@@ -53,32 +45,5 @@
         "node": ">=0.10.0"
       }
     }
-  },
-  "dependencies": {
-    "nanoid": {
-      "version": "3.3.1",
-      "resolved": "https://registry.npmjs.org/nanoid/-/nanoid-3.3.1.tgz",
-      "integrity": "sha512-n6Vs/3KGyxPQd6uO0eH4Bv0ojGSUvuLlIHtC3Y0kEO23YRge8H9x1GCzLn28YX0H66pMkxuaeESFq4tKISKwdw=="
-    },
-    "picocolors": {
-      "version": "1.0.0",
-      "resolved": "https://registry.npmjs.org/picocolors/-/picocolors-1.0.0.tgz",
-      "integrity": "sha512-1fygroTLlHu66zi26VoTDv8yRgm0Fccecssto+MhsZ0D/DGW2sm8E8AjW7NU5VVTRt5GxbeZ5qBuJr+HyLYkjQ=="
-    },
-    "postcss": {
-      "version": "8.4.7",
-      "resolved": "https://registry.npmjs.org/postcss/-/postcss-8.4.7.tgz",
-      "integrity": "sha512-L9Ye3r6hkkCeOETQX6iOaWZgjp3LL6Lpqm6EtgbKrgqGGteRMNb9vzBfRL96YOSu8o7x3MfIH9Mo5cPJFGrW6A==",
-      "requires": {
-        "nanoid": "^3.3.1",
-        "picocolors": "^1.0.0",
-        "source-map-js": "^1.0.2"
-      }
-    },
-    "source-map-js": {
-      "version": "1.0.2",
-      "resolved": "https://registry.npmjs.org/source-map-js/-/source-map-js-1.0.2.tgz",
-      "integrity": "sha512-R0XvVJ9WusLiqTCEiGCmICCMplcCkIwwR11mOSD9CR5u+IXYdiseeEuXCVAjS54zqwkLcPNnmU4OeJ6tUrWhDw=="
-    }
   }
 }
```

And the full allegedly version 3 lock file is
```sh
cat node_modules/.package-lock.json
{
  "name": "renovate-npm-lockfile-diff-repro",
  "version": "1.0.0",
  "lockfileVersion": 2,
  "requires": true,
  "packages": {
    "node_modules/nanoid": {
      "version": "3.3.1",
      "resolved": "https://registry.npmjs.org/nanoid/-/nanoid-3.3.1.tgz",
      "integrity": "sha512-n6Vs/3KGyxPQd6uO0eH4Bv0ojGSUvuLlIHtC3Y0kEO23YRge8H9x1GCzLn28YX0H66pMkxuaeESFq4tKISKwdw==",
      "bin": {
        "nanoid": "bin/nanoid.cjs"
      },
      "engines": {
        "node": "^10 || ^12 || ^13.7 || ^14 || >=15.0.1"
      }
    },
    "node_modules/picocolors": {
      "version": "1.0.0",
      "resolved": "https://registry.npmjs.org/picocolors/-/picocolors-1.0.0.tgz",
      "integrity": "sha512-1fygroTLlHu66zi26VoTDv8yRgm0Fccecssto+MhsZ0D/DGW2sm8E8AjW7NU5VVTRt5GxbeZ5qBuJr+HyLYkjQ=="
    },
    "node_modules/postcss": {
      "version": "8.4.7",
      "resolved": "https://registry.npmjs.org/postcss/-/postcss-8.4.7.tgz",
      "integrity": "sha512-L9Ye3r6hkkCeOETQX6iOaWZgjp3LL6Lpqm6EtgbKrgqGGteRMNb9vzBfRL96YOSu8o7x3MfIH9Mo5cPJFGrW6A==",
      "dependencies": {
        "nanoid": "^3.3.1",
        "picocolors": "^1.0.0",
        "source-map-js": "^1.0.2"
      },
      "engines": {
        "node": "^10 || ^12 || >=14"
      },
      "funding": {
        "type": "opencollective",
        "url": "https://opencollective.com/postcss/"
      }
    },
    "node_modules/source-map-js": {
      "version": "1.0.2",
      "resolved": "https://registry.npmjs.org/source-map-js/-/source-map-js-1.0.2.tgz",
      "integrity": "sha512-R0XvVJ9WusLiqTCEiGCmICCMplcCkIwwR11mOSD9CR5u+IXYdiseeEuXCVAjS54zqwkLcPNnmU4OeJ6tUrWhDw==",
      "engines": {
        "node": ">=0.10.0"
      }
    }
  }
}
```

So, there are just a couple of things missing from the version 2 lock file. And one of them being the block of information that is in the v2 lock file, that I want to update. That information is also not existing in the v1 lock file. => it should be fine to only handle v2 here to my understanding.

**Comment by herndlm** [comment]
> @herndlm do you have any public repo you can demo this against, when done?

yes, of course. the one I reproduced this with. and for sake of completeness I added another lockfileVersion 1 repo too, which shows that it is not affected by the problem.

https://github.com/herndlm/renovate-npm-lockfile-diff-repro/pull/3
https://github.com/herndlm/renovate-npm-lockfile-diff-repro-v1/pull/3

**Comment by herndlm** [comment]
just wondering if the PR title should be changed to make it easier to understand what happens here, but feel free to use whatever you prefer :)

**Comment by herndlm** [comment]
@rarkins sorry, I might have misunderstood you, do you want a demo in this state before merging, or after?

**Review by rarkins — APPROVED** [review]

**Review by rarkins — CHANGES_REQUESTED** [review]

**Review by rarkins — APPROVED** [review]

**Comment by herndlm** [comment]
Sorry, didn't want to rush things here, I can wait :)

**Comment by herndlm** [comment]
This is fixing the issue, I'm running 32.6.2 internally already. But, I saw that there seems to be a newline missing in the end of package-lock.json 🤦‍♂️ I'm going to double-check and try to open a fix PR later. Maybe I'll do a small refactor of the lockfile parsing + saving first and move parts into util functions. Because it's already done in 2 places now and I assume both are affected.

**Pull request merged**